Move ocpp.FormationViolation to Endpoint (BC)

andig opened the pull request:
Fix https://github.com/lorenzodonini/ocpp-go/issues/227

TODO

- [x] fix tests

lorenzodonini commented:
> The ocppj tests look wrong to me- since they never setup a server/client they cannot validate against endpoint's format.

What do you mean? Server/client are created for each test within the `SetupTest` function. Only the websocket layer is mocked.

What tests are broken exactly? Happy to help and think of something.

lorenzodonini commented on ocppj/ocppj.go:
```diff
@@ -198,10 +198,6 @@ const (
-	FormationViolation = FormatViolationV16 // Used as constant, but can be overwritten depending on protocol version. Sett FormatViolationV16 and FormatViolationV2.
-)
-
 func IsErrorCodeValid(fl validator.FieldLevel) bool {
```
By removing the global `FormationViolation` variable this validation function will not work as expected, since the switch case below checks for this value.

I see two options:
1. move the validation function to the 1.6/2.0 packages respectively
  - cleaner but will not allow using 1.6+2.0 within the same process (something which is apprently a requirement for some devs)
2. replace `FormationViolation` in the switch case with `FormatViolationV16, FormatViolationV2`
  - works with both proto versions at the same time but doesn't prevent other implementations from sending an "incorrect" value for a specific version

I'd honestly go with 2, cause it's easier and will make everyone happy in the end.

andig replied:
> By removing the global FormationViolation variable this validation function will not work as expected, since the switch case below checks for this value.

Doh: I see. It could work if the validation could get access to the endpoint- but I guess that's not something that the validator would allow hooking in (something like a validation context)?

> move the validation function to the 1.6/2.0 packages respectively

It would not help either since clients and servers would still race.

> replace FormationViolation in the switch case with FormatViolationV16, FormatViolationV2

I can't judge but it seems nice and simple and still better than a data race?

andig commented:
> What do you mean? Server/client are created for each test within the SetupTest function. Only the websocket layer is mocked.

Afaikt they create mock client/servers. Missing the endpoint they cannot be used to test for the endpoint's format.

lorenzodonini commented:
All the failing tests are checking against the removed global var. To fix them, I suggest simply changing `ocppj.FormationViolation` into respectively:
- `suite.chargePoint.Endpoint.FormatError` or
- `suite.centralSystem.Endpoint.FormatError`

andig commented:
Thank you, updated accordingly. That fixes it for ocppj. for ocpp16 and 20 the constructors are private and the tests are in a separate folder. Any idea how to tackle this?

dwibudut commented:
Hello,
I upvote for this changes. With this change will not overwrite the `ocppj.FormationViolation` global value with the last one called (when start server each version). And I think for the fix test error need to split `ocppj` test for each version too (1.6 and 2.0.1)

lorenzodonini commented on ocpp2.0.1/csms.go:
```diff
@@ -811,11 +811,15 @@ func (cs *csms) SendRequestAsync(clientId string, request ocpp.Request, callback
 func (cs *csms) Start(listenPort int, listenPath string) {
 	// Overriding some protocol-specific values in the lower layers globally
-	ocppj.FormationViolation = ocppj.FormatViolationV2
+	cs.server.Endpoint.Dialect = ocpp.V16
```
I guess this should be:

```suggestion
	cs.server.Endpoint.Dialect = ocpp.V2
```

andig commented:
Tests currently panic since I have tried removing the default version for the test suite:

	// defaultDialect := ocpp.V16 // set default to version 1.6 format error *for test only
	// suite.centralSystem.Dialect = defaultDialect
	// suite.chargePoint.Dialect = defaultDialect

Tbo, it is not clear for me why there should be a default version. Which default version should an endpoint have when it is being constructed? This feels like some sort of logical gap that we might fix while we're at it.

@lorenzodonini Is this a logical error in where the endpoint dialect (which is what I'm currently calling it) is being set or is this rather a mistake in the tests that should not test that specific property in these tests?

dwibudut commented:
Error caused by dialect that not set to any version, and then the dialect value is 0, that it's made panic on `ocppj.FormatErrorForDialect`.

Suppose when `ocppj.FormatErrorForDialect` set default to returned empty string, this will error too in `ErrorCode` value validator.

Have you like to traced it with go test for first error method test, like:

```
go test -v ./ocppj -run ^TestMockOcppJ$ -testify.m TestCentralSystemInvalidMessageHook
```

dwibudut commented:
To avoid mistake or logical gap, prefer need to make two different test version between:
```
defaultDialect := ocpp.V16
```

and

```
defaultDialect := ocpp.V2
```

This might be fairly test.

andig commented:
Finished from my side. Didn‘t get rid of the test default but that should be acceptable.

lorenzodonini commented on ocpp1.6/charge_point.go:
```diff
@@ -334,7 +334,7 @@ func (cp *chargePoint) sendResponse(confirmation ocpp.Response, err error, reque
 func (cp *chargePoint) Start(centralSystemUrl string) error {
 	// Overriding some protocol-specific values in the lower layers globally
-	ocppj.FormationViolation = ocppj.FormatViolationV16
+	cp.client.Endpoint.SetDialect(ocpp.V16)
```
I would honestly move this logic to the `ocpp16.NewChargePoint` constructor function (respectively for both endpoints for both proto versions). It's a static setting that should be set at creation time, along with the sub-protocol and base handlers.

andig replied:
Brilliant. Thought about that but could not finde the right place to do so.

lorenzodonini commented on ocppj/ocppj_test.go:
```diff
@@ -375,6 +374,9 @@ func (suite *OcppJTestSuite) SetupTest() {
 	suite.centralSystem = ocppj.NewServer(suite.mockServer, suite.serverDispatcher, nil, mockProfile)
+	defaultDialect := ocpp.V16 // set default to version 1.6 format error *for test only
+	suite.centralSystem.SetDialect(defaultDialect)
+	suite.chargePoint.SetDialect(defaultDialect)
```
> Didn‘t get rid of the test default but that should be acceptable.

These will sadly still be required, but not a big deal.

lorenzodonini commented on ocpp1.6_test/proto_test.go:
```diff
@@ -157,5 +157,5 @@ func (suite *OcppV16TestSuite) TestErrorCodes() {
 	suite.mockWsServer.On("Start", mock.AnythingOfType("int"), mock.AnythingOfType("string")).Return(nil)
 	suite.centralSystem.Start(8887, "somePath")
-	assert.Equal(t, ocppj.FormatViolationV16, ocppj.FormationViolation)
+	assert.Equal(t, ocppj.FormatViolationV16, ocppj.FormatErrorType(suite.centralSystem))
```
Wouldn't it be fine to just run the function against `suite.ocppjClient` (or ocppjServer)? These also implement the `dialector` interface, so it would save one more function in the v16/v2 package. All tests should have access to the ocppj structs directly.

andig replied:
Brilliant- exactly what I wanted!

andig commented:
@lorenzodonini addressed comments- done. I like it, much appreciated!

Small comment: note the calls to `suite.Equal`- there is some potential to simplify the test code a bit.
